etc/run-clang-tidy-cached.cc

```diff
@@ -0,0 +1,627 @@
+        clang_tidy_args_(AssembleArgs(argc, argv))
+  {
+    project_cache_dir_ = AssembleProjectCacheDir(cache_prefix);
+  }
+
+  const fs::path& project_cache_dir() const { return project_cache_dir_; }
+
+  // Given a work-queue in/out-file, process it. Using system() for portability.
+  // Empties work_queue.
+  void RunClangTidyOn(ContentAddressedStore& output_store,
+                      std::list<filepath_contenthash_t>* work_queue)
+  {
+    if (work_queue->empty()) {
+      return;
+    }
+    const char* jobs_env_str = getenv("CLANG_TIDY_JOBS");
+    const int jobs_env_num = jobs_env_str ? atoi(jobs_env_str) : -1;
+    const int kJobs = (jobs_env_num > 0 ? jobs_env_num
+                                        : std::thread::hardware_concurrency());
+    std::cerr << work_queue->size() << " files to process (w/ " << kJobs
+              << " jobs)...";
+
+    const bool print_progress = isatty(STDERR_FILENO);
+    if (!print_progress) {
+      std::cerr << "\n";
+    }
+
+    std::mutex queue_access_lock;
+    auto clang_tidy_runner = [&]() {
+      for (;;) {
+        filepath_contenthash_t work;
+        {
+          const std::lock_guard<std::mutex> lock(queue_access_lock);
+          if (work_queue->empty()) {
+            return;
+          }
+          if (print_progress) {
+            fprintf(stderr, "%5d\b\b\b\b\b", (int) (work_queue->size()));
+          }
+          work = work_queue->front();
+          work_queue->pop_front();
+        }
+        const fs::path final_out = output_store.PathFor(work);
+        const std::string tmp_out = final_out.string() + ".tmp";
+        // Putting the file to clang-tidy early in the command line so that
+        // it is easy to find with `ps` or `top`.
+        const std::string command = clang_tidy_ + " '" + work.first.string()
+                                    + "'" + clang_tidy_args_ + "> '" + tmp_out
+                                    + "' 2>/dev/null";
+        const int r = system(command.c_str());
+#ifdef WIFSIGNALED
+        // NOLINTBEGIN
+        if (WIFSIGNALED(r)
+            && (WTERMSIG(r) == SIGINT || WTERMSIG(r) == SIGQUIT)) {
+          break;  // got Ctrl-C
+        }
+        // NOLINTEND
+#endif
+        RepairFilenameOccurences(tmp_out, tmp_out);
+        fs::rename(tmp_out, final_out);  // atomic replacement
+      }
+    };
+
+    std::vector<std::thread> workers;
+    for (auto i = 0; i < kJobs; ++i) {
+      workers.emplace_back(clang_tidy_runner);  // NOLINT
+    }
+    for (auto& t : workers) {
+      t.join();
+    }
+    if (print_progress) {
+      fprintf(stderr, "     \n");  // Clean out progress counter.
+    }
+  }
+
+ private:
+  static fs::path GetCacheBaseDir()
+  {
+    if (const char* from_env = getenv("CACHE_DIR")) {
+      return fs::path{from_env};
+    }
+    if (const char* home = getenv("HOME")) {
+      if (auto cdir = fs::path(home) / ".cache/"; fs::exists(cdir)) {
+        return cdir;
+      }
+    }
```

Comment:
If you cache in the home dir you may run into conflicts between runs in different work areas.  Is there a policy to remove old cache entries or does it grow indefinately?

Reply:
the cache is separated by project and hash of the clang-tidy configuration used, so it will not clash with other projects. It will share the cache between different checkouts of OpenRoad, but that is desired, as it allows to quickly switch between branches and re-create the output almost instantaneously.

Old cache entries are currently not automatically removed (it could be added but for now this is not a large problem in particular if the project is mostly clang-tidy clean (which we're not quite yet :) )).

Reply:
What guards against two different concurrent runs of OR in different areas colliding in the shared cache?

Reply:
The outputs are separate files that are content-addressed, so their name depends on the hash of its content and the headers it depends on (and the clang-tidy configuration). Which means that even if there are two instances in parallel running on exactly the same file they will create the same content anyway. That writes into a temporary file.

After that, if the file is fully created without issues, it is atomically moved to its final place. So even if there were two instances running, there is only one 'winner' (which then serves both).